// nuclide-server/lib/NuclideRemoteEventbus.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }; })();

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _nuclideCommons = require('nuclide-commons');

var _serviceframework = require('./serviceframework');

var _serviceframework2 = _interopRequireDefault(_serviceframework);

'use babel';

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _require = require('event-kit');

var Disposable = _require.Disposable;

var _require2 = require('./service-manager');

var getRemoteEventName = _require2.getRemoteEventName;

var _require3 = require('./utils');

var serializeArgs = _require3.serializeArgs;

var _require4 = require('events');

var EventEmitter = _require4.EventEmitter;

var NuclideSocket = require('./NuclideSocket');

var _require5 = require('./config');

var SERVICE_FRAMEWORK_EVENT_CHANNEL = _require5.SERVICE_FRAMEWORK_EVENT_CHANNEL;
var SERVICE_FRAMEWORK_RPC_CHANNEL = _require5.SERVICE_FRAMEWORK_RPC_CHANNEL;
var SERVICE_FRAMEWORK_RPC_TIMEOUT_MS = _require5.SERVICE_FRAMEWORK_RPC_TIMEOUT_MS;
var SERVICE_FRAMEWORK3_CHANNEL = _require5.SERVICE_FRAMEWORK3_CHANNEL;

var logger = require('nuclide-logging').getLogger();

var NuclideRemoteEventbus = (function () {
  function NuclideRemoteEventbus(serverUri) {
    var _this = this;

    var options = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    _classCallCheck(this, NuclideRemoteEventbus);

    this.socket = new NuclideSocket(serverUri, options);
    this.socket.on('message', function (message) {
      return _this._handleSocketMessage(message);
    });
    this.serviceFrameworkEventEmitter = new EventEmitter();
    this._rpcRequestId = 1;
    this._serviceFrameworkRpcEmitter = new EventEmitter();
    this._serviceFramework3Emitter = new EventEmitter();

    this._clientComponent = new _serviceframework2['default'].ClientComponent(this._serviceFramework3Emitter, this.socket, function () {
      return _this._rpcRequestId++;
    });
  }

  // Resolves if the connection looks healthy.
  // Will reject quickly if the connection looks unhealthy.

  _createClass(NuclideRemoteEventbus, [{
    key: 'testConnection',
    value: function testConnection() {
      // Don't call this after socket is closed.
      (0, _assert2['default'])(this.socket != null);
      return this.socket.testConnection();
    }
  }, {
    key: '_handleSocketMessage',
    value: function _handleSocketMessage(message) {
      var channel = message.channel;
      var event = message.event;

      if (channel === SERVICE_FRAMEWORK_RPC_CHANNEL) {
        var requestId = message.requestId;
        var error = message.error;
        var result = message.result;

        this._serviceFrameworkRpcEmitter.emit(requestId.toString(), error, result);
        return;
      }

      if (channel === SERVICE_FRAMEWORK_EVENT_CHANNEL) {
        this.serviceFrameworkEventEmitter.emit.apply(this.serviceFrameworkEventEmitter, [event.name].concat(event.args));
        return;
      }

      if (channel === SERVICE_FRAMEWORK3_CHANNEL) {
        var requestId = message.requestId;
        var hadError = message.hadError;
        var error = message.error;
        var result = message.result;

        this._serviceFramework3Emitter.emit(requestId.toString(), hadError, error, result);
        return;
      }
    }
  }, {
    key: '_subscribeEventOnServer',
    value: function _subscribeEventOnServer(serviceName, methodName, serviceOptions) {
      return this.callServiceFrameworkMethod('serviceFramework', 'subscribeEvent',
      /*methodArgs*/[this.socket.id, serviceName, methodName], serviceOptions);
    }
  }, {
    key: '_unsubscribeEventFromServer',
    value: function _unsubscribeEventFromServer(serviceName, methodName, serviceOptions) {
      return this.callServiceFrameworkMethod('serviceFramework', 'unsubscribeEvent',
      /*methodArgs*/[this.socket.id, serviceName, methodName], serviceOptions);
    }
  }, {
    key: 'registerEventListener',
    value: function registerEventListener(localEventName, callback, serviceOptions) {
      var _this2 = this;

      var _localEventName$split = localEventName.split('/');

      var _localEventName$split2 = _slicedToArray(_localEventName$split, 2);

      var serviceName = _localEventName$split2[0];
      var eventMethodName = _localEventName$split2[1];

      var remoteEventName = getRemoteEventName(serviceName, eventMethodName, serviceOptions);
      this.serviceFrameworkEventEmitter.on(remoteEventName, callback);
      var subscribePromise = this._subscribeEventOnServer(serviceName, eventMethodName, serviceOptions);
      return new Disposable(function () {
        _this2.serviceFrameworkEventEmitter.removeListener(remoteEventName, callback);
        return subscribePromise.then(function () {
          return _this2._unsubscribeEventFromServer(serviceName, eventMethodName, serviceOptions);
        });
      });
    }
  }, {
    key: 'callMethod',
    value: _asyncToGenerator(function* (serviceName, methodName, methodArgs, extraOptions) {
      if (!this.socket) {
        logger.error('RemoteEventBus closed - callMethod:', serviceName, methodName);
        // Error condition that should never happen, return `undefined`.
        return;
      }

      var _serializeArgs = serializeArgs(methodArgs || []);

      var args = _serializeArgs.args;
      var argTypes = _serializeArgs.argTypes;

      try {
        return yield this.socket.xhrRequest(_nuclideCommons.object.assign({
          uri: serviceName + '/' + methodName,
          qs: {
            args: args,
            argTypes: argTypes
          },
          method: 'GET' }, // default request method is 'GET'.
        extraOptions || {}));
      } catch (err) {
        logger.error(err);
        throw err;
      }
    })
  }, {
    key: 'callServiceFrameworkMethod',
    value: _asyncToGenerator(function* (serviceName, methodName, methodArgs, serviceOptions) {
      var _this3 = this;

      var timeout = arguments.length <= 4 || arguments[4] === undefined ? SERVICE_FRAMEWORK_RPC_TIMEOUT_MS : arguments[4];

      var requestId = this._rpcRequestId++;

      this.socket.send({
        serviceName: serviceName,
        methodName: methodName,
        methodArgs: methodArgs,
        serviceOptions: serviceOptions,
        requestId: requestId
      });

      return new Promise(function (resolve, reject) {
        _this3._serviceFrameworkRpcEmitter.once(requestId.toString(), function (error, result) {
          error ? reject(error) : resolve(result);
        });

        setTimeout(function () {
          _this3._serviceFrameworkRpcEmitter.removeAllListeners(requestId.toString());
          reject('Timeout after ' + timeout + ' for ' + serviceName + '/' + methodName);
        }, timeout);
      });
    })

    // Delegate RPC functions to ServiceFramework.ClientComponent
  }, {
    key: 'callRemoteFunction',
    value: function callRemoteFunction() {
      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return this._clientComponent.callRemoteFunction.apply(this._clientComponent, args);
    }
  }, {
    key: 'createRemoteObject',
    value: function createRemoteObject() {
      for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return this._clientComponent.createRemoteObject.apply(this._clientComponent, args);
    }
  }, {
    key: 'callRemoteMethod',
    value: function callRemoteMethod() {
      for (var _len3 = arguments.length, args = Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
        args[_key3] = arguments[_key3];
      }

      return this._clientComponent.callRemoteMethod.apply(this._clientComponent, args);
    }
  }, {
    key: 'disposeRemoteObject',
    value: function disposeRemoteObject() {
      for (var _len4 = arguments.length, args = Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
        args[_key4] = arguments[_key4];
      }

      return this._clientComponent.disposeRemoteObject.apply(this._clientComponent, args);
    }

    // Delegate marshalling to the ServiceFramework.ClientComponent class.
  }, {
    key: 'marshal',
    value: function marshal() {
      var _clientComponent;

      return (_clientComponent = this._clientComponent).marshal.apply(_clientComponent, arguments);
    }
  }, {
    key: 'unmarshal',
    value: function unmarshal() {
      var _clientComponent2;

      return (_clientComponent2 = this._clientComponent).unmarshal.apply(_clientComponent2, arguments);
    }
  }, {
    key: 'registerType',
    value: function registerType() {
      var _clientComponent3;

      return (_clientComponent3 = this._clientComponent).registerType.apply(_clientComponent3, arguments);
    }
  }, {
    key: 'close',
    value: function close() {
      this.socket.close();
      this.socket = null;
    }
  }]);

  return NuclideRemoteEventbus;
})();

module.exports = NuclideRemoteEventbus;
