// nuclide-debugger-hhvm-proxy/lib/properties.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _require = require('./utils');

var log = _require.log;

var _require2 = require('./ObjectId');

var remoteObjectIdOfObjectId = _require2.remoteObjectIdOfObjectId;
var endIndexOfObjectId = _require2.endIndexOfObjectId;
var startIndexOfObjectId = _require2.startIndexOfObjectId;
var countOfObjectId = _require2.countOfObjectId;
var getChildIds = _require2.getChildIds;

var _require3 = require('./values');

var convertValue = _require3.convertValue;

function convertProperties(id, properties) {
  log('Got properties: ' + JSON.stringify(properties));
  return properties.map(function (property) {
    return convertProperty(id, property);
  });
}

/**
 * Converts a DbgpProperty to a Chrome PropertyDescriptor.
 */
function convertProperty(contextId, dbgpProperty) {
  log('Converting to Chrome property: ' + JSON.stringify(dbgpProperty));
  var result = {
    configurable: false,
    enumerable: true,
    name: dbgpProperty.$.name,
    value: convertValue(contextId, dbgpProperty)
  };
  return result;
}

/**
 * Given an ObjectId for a multi page object, gets PropertyDescriptors
 * for the object's children.
 */
function getPagedProperties(pagedId) {
  var pagesize = pagedId.elementRange.pagesize;
  var endIndex = endIndexOfObjectId(pagedId);

  var childIds = getChildIds(pagedId);
  return childIds.map(function (childId) {
    var childStartIndex = startIndexOfObjectId(childId, pagesize);
    var childCount = countOfObjectId(childId, pagesize, endIndex);
    return {
      configurable: false,
      enumerable: true,
      name: 'Elements(' + childStartIndex + '..' + (childStartIndex + childCount - 1) + ')',
      value: {
        description: childCount + ' elements',
        type: 'object',
        objectId: remoteObjectIdOfObjectId(childId)
      }
    };
  });
}

module.exports = {
  convertProperties: convertProperties,
  convertProperty: convertProperty,
  getPagedProperties: getPagedProperties
};
