// nuclide-clang/lib/LocalClangService.js
var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _findClangServerArgs = _asyncToGenerator(function* () {
  var findClangServerArgs = undefined;
  try {
    findClangServerArgs = require('./fb/find-clang-server-args');
  } catch (e) {
    // Ignore.
  }

  var libClangLibraryFile = undefined;
  if (process.platform === 'darwin') {
    var result = yield checkOutput('xcode-select', ['--print-path']);
    if (result.exitCode === 0) {
      libClangLibraryFile = result.stdout.trim() + '/Toolchains/XcodeDefault.xctoolchain/usr/lib/libclang.dylib';
    }
  }

  var clangServerArgs = {
    libClangLibraryFile: libClangLibraryFile,
    pythonExecutable: 'python',
    pythonPathEnv: path.join(__dirname, '../pythonpath')
  };
  if (typeof findClangServerArgs === 'function') {
    var clangServerArgsOverrides = yield findClangServerArgs();
    return _extends({}, clangServerArgs, clangServerArgsOverrides);
  } else {
    return clangServerArgs;
  }
});

var createAsyncConnection = _asyncToGenerator(function* (pathToLibClangServer) {
  // $FlowIssue D2268946
  return yield new Promise(_asyncToGenerator(function* (resolve, reject) {
    var _ref = yield _findClangServerArgs();

    var libClangLibraryFile = _ref.libClangLibraryFile;
    var pythonPathEnv = _ref.pythonPathEnv;
    var pythonExecutable = _ref.pythonExecutable;

    var options = {
      cwd: path.dirname(pathToLibClangServer),
      // The process should use its ordinary stderr for errors.
      stdio: ['pipe', null, 'pipe', 'pipe'],
      detached: false, // When Atom is killed, clang_server.py should be killed, too.
      env: {
        // On Mac OSX El Capitan, bash seems to wipe out the `LD_LIBRARY_PATH` and
        // `DYLD_LIBRARY_PATH` environment variables. So, set this env var which is read by
        // clang_server.py to explicitly set the file path to load.
        LIB_CLANG_LIBRARY_FILE: libClangLibraryFile,
        PYTHONPATH: pythonPathEnv
      }
    };

    // Note that safeSpawn() often overrides options.env.PATH, but that only happens when
    // options.env is undefined (which is not the case here). This will only be an issue if the
    // system cannot find `pythonExecutable`.
    var child = yield safeSpawn(pythonExecutable, /* args */[pathToLibClangServer], options);
    child.on('close', function (exitCode) {
      logger.error('%s exited with code %s', pathToLibClangServer, exitCode);
    });
    child.stderr.on('data', function (error) {
      if (error instanceof Buffer) {
        error = error.toString('utf8');
      }
      logger.error('Error receiving data', error);
    });
    /* $FlowFixMe - update Flow defs for ChildProcess */
    var writableStream = child.stdio[3];

    // Make sure the bidirectional communication channel is set up before
    // resolving this Promise.
    child.stdout.once('data', function (data) {
      if (data.toString() === 'ack\n') {
        var result = {
          readableStream: child.stdout,
          writableStream: writableStream
        };
        resolve(result);
      } else {
        reject(data);
      }
    });
    writableStream.write('init\n');
  }));
}

/**
 * Nuclide runs clang in its own process. For simplicity, it is run as a Python
 * program, as Python has a nice libclang wrapper.
 */
);

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _nuclideBuckBaseLibBuckUtils = require('nuclide-buck-base/lib/BuckUtils');

'use babel';

var _require = require('nuclide-commons');

var checkOutput = _require.checkOutput;
var fsPromise = _require.fsPromise;
var safeSpawn = _require.safeSpawn;

var logger = require('nuclide-logging').getLogger();
var path = require('path');
var split = require('split');

var _require2 = require('events');

var EventEmitter = _require2.EventEmitter;

var ClangFlagsManager = require('./ClangFlagsManager');
var ClangService = require('./ClangService');

var LocalClangService = (function (_ClangService) {
  _inherits(LocalClangService, _ClangService);

  function LocalClangService() {
    _classCallCheck(this, LocalClangService);

    _get(Object.getPrototypeOf(LocalClangService.prototype), 'constructor', this).call(this);
    this._asyncConnection = null;
    this._clangFlagsManager = new ClangFlagsManager(new _nuclideBuckBaseLibBuckUtils.BuckUtils());
    this._readableStream = null;
    this._writableStream = null;
    this._nextRequestId = 1;
    this._pathToLibClangServer = path.join(__dirname, '../python/clang_server.py');
    this._emitter = new EventEmitter();
  }

  _createClass(LocalClangService, [{
    key: '_connect',
    value: function _connect() {
      var _this = this;

      var asyncConnection = this._asyncConnection;
      if (asyncConnection == null) {
        asyncConnection = createAsyncConnection(this._pathToLibClangServer).then(function (connection) {
          var readableStream = connection.readableStream;
          var writableStream = connection.writableStream;

          _this._readableStream = readableStream;
          _this._writableStream = writableStream;
          readableStream.pipe(split(JSON.parse)).on('data', function (response) {
            var id = response['reqid'];
            _this._emitter.emit(id, response);
          });
        }, function (error) {
          // If an error occurs, clear out `this._asyncConnection`, so that if `_connect()` is
          // called again, we make a new attempt to create a connection,
          // rather than holding onto a rejected Promise indefinitely.
          _this._asyncConnection = null;
        });
      }
      this._asyncConnection = asyncConnection;
      return asyncConnection;
    }
  }, {
    key: '_makeRequest',
    value: _asyncToGenerator(function* (request) {
      var _this2 = this;

      yield this._connect();

      var id = request['reqid'] = this._getNextRequestId();
      var logData = JSON.stringify(request, function (key, value) {
        // File contents are too large and clutter up the logs, so exclude them.
        if (key === 'contents') {
          return undefined;
        } else {
          return value;
        }
      });
      logger.debug('LibClang request: ' + logData);
      var writableStream = this._writableStream;
      if (writableStream == null) {
        return;
      }
      // Because Node uses an event-loop, we do not have to worry about a call to
      // write() coming in from another thread between our two calls here.
      writableStream.write(JSON.stringify(request));
      writableStream.write('\n');

      return new Promise(function (resolve, reject) {
        _this2._emitter.once(id, function (response) {
          var isError = ('error' in response);
          if (isError) {
            logger.error('error received from clang_server.py for request:\n%o\nError:%s', logData, response['error']);
          }
          (isError ? reject : resolve)(response);
        });
      });
    })
  }, {
    key: '_getNextRequestId',
    value: function _getNextRequestId() {
      return (this._nextRequestId++).toString(16);
    }
  }, {
    key: 'compile',
    value: _asyncToGenerator(function* (src, contents) {
      var flags = yield this._clangFlagsManager.getFlagsForSrc(src);
      return this._makeRequest({
        method: 'compile',
        src: src,
        contents: contents,
        flags: flags
      });
    })
  }, {
    key: 'getCompletions',
    value: _asyncToGenerator(function* (src, contents, line, column, tokenStartColumn, prefix) {
      var flags = yield this._clangFlagsManager.getFlagsForSrc(src);
      return this._makeRequest({
        method: 'get_completions',
        src: src,
        contents: contents,
        line: line,
        column: column,
        tokenStartColumn: tokenStartColumn,
        prefix: prefix,
        flags: flags
      });
    })

    /**
     * @return An Object with the following fields:
     *   file: NuclideUri of the source file where the declaration is.
     *   line: The line number of the declaration within `file`.
     *   column: The column number of the declaration within `file`.
     *   spelling: The spelling of the entity.
     */
  }, {
    key: 'getDeclaration',
    value: _asyncToGenerator(function* (src, contents, line, column) {
      var flags = yield this._clangFlagsManager.getFlagsForSrc(src);
      var data = yield this._makeRequest({
        method: 'get_declaration',
        src: src,
        contents: contents,
        line: line,
        column: column,
        flags: flags
      });

      var locationAndSpelling = data.locationAndSpelling;

      if (!locationAndSpelling) {
        return null;
      }

      var state = yield fsPromise.lstat(locationAndSpelling.file);
      if (state.isSymbolicLink()) {
        locationAndSpelling.file = yield fsPromise.readlink(locationAndSpelling.file);
      }

      return locationAndSpelling;
    })
  }, {
    key: 'getDeclarationInfo',
    value: _asyncToGenerator(function* (src, contents, line, column) {
      var flags = yield this._clangFlagsManager.getFlagsForSrc(src);
      return this._makeRequest({
        method: 'get_declaration_info',
        src: src,
        contents: contents,
        line: line,
        column: column,
        flags: flags
      });
    })
  }]);

  return LocalClangService;
})(ClangService);

module.exports = LocalClangService;
