// nuclide-server/lib/serviceframework/ServerComponent.js
Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _rx = require('rx');

var _nuclideServiceParser = require('nuclide-service-parser');

var _config = require('./config');

var _NuclideServer = require('../NuclideServer');

var _NuclideServer2 = _interopRequireDefault(_NuclideServer);

var _nuclideServiceParserLibTypeRegistry = require('nuclide-service-parser/lib/TypeRegistry');

var _nuclideServiceParserLibTypeRegistry2 = _interopRequireDefault(_nuclideServiceParserLibTypeRegistry);

'use babel';

var logger = require('nuclide-logging').getLogger();

var ServerComponent = (function () {
  function ServerComponent(server) {
    var _this = this;

    _classCallCheck(this, ServerComponent);

    this._server = server;

    this._typeRegistry = new _nuclideServiceParserLibTypeRegistry2['default']();
    this._functionsByName = new Map();
    this._classesByName = new Map();

    this._nextObjectId = 1;
    this._objectRegistry = new Map();

    this._subscriptions = new Map();

    // NuclideUri type requires no transformations (it is done on the client side).
    this._typeRegistry.registerType('NuclideUri', function (uri) {
      return uri;
    }, function (remotePath) {
      return remotePath;
    });

    var services = (0, _config.loadServicesConfig)();

    var _loop = function (service) {
      logger.debug('Registering 3.0 service ' + service.name + '...');
      try {
        (function () {
          var defs = (0, _nuclideServiceParser.getDefinitions)(service.definition);
          var localImpl = require(service.implementation);

          // Register type aliases.
          defs.forEach(function (definition) {
            var name = definition.name;
            switch (definition.kind) {
              case 'alias':
                logger.debug('Registering type alias ' + name + '...');
                if (definition.definition != null) {
                  _this._typeRegistry.registerAlias(name, definition.definition);
                }
                break;
              case 'function':
                // Register module-level functions.
                _this._registerFunction(service.name + '/' + name, localImpl[name], definition.type);
                break;
              case 'interface':
                // Register interfaces.
                logger.debug('Registering interface ' + name + '...');
                _this._classesByName.set(name, {
                  localImplementation: localImpl[name],
                  definition: definition
                });

                _this._typeRegistry.registerType(name, _asyncToGenerator(function* (object) {
                  // If the object has already been assigned an id, return that id.
                  if (object._remoteId) {
                    return object._remoteId;
                  }

                  // Put the object in the registry.
                  object._interface = name;
                  var objectId = _this._nextObjectId;
                  _this._objectRegistry.set(objectId, object);
                  object._remoteId = objectId;
                  _this._nextObjectId++;

                  return objectId;
                }), _asyncToGenerator(function* (objectId) {
                  return _this._objectRegistry.get(objectId);
                }));

                // Register all of the static methods as remote functions.
                definition.staticMethods.forEach(function (funcType, funcName) {
                  _this._registerFunction(name + '/' + funcName, localImpl[name][funcName], funcType);
                });
                break;
            }
          });
        })();
      } catch (e) {
        logger.error('Failed to load service ' + service.name + '. Stack Trace:\n' + e.stack);
        return 'continue';
      }
    };

    for (var service of services) {
      var _ret = _loop(service);

      if (_ret === 'continue') continue;
    }
  }

  /**
   * A helper function that let's us 'apply' an array of arguments to a constructor.
   * It works by creating a new constructor that has the same prototype as the original
   * constructor, and simply applies the original constructor directly to 'this'.
   * @returns An instance of classObject.
   */

  _createClass(ServerComponent, [{
    key: '_registerFunction',
    value: function _registerFunction(name, localImpl, type) {
      logger.debug('Registering function ' + name + '...');
      if (this._functionsByName.has(name)) {
        throw new Error('Duplicate RPC function: ' + name);
      }
      this._functionsByName.set(name, {
        localImplementation: localImpl,
        type: type
      });
    }
  }, {
    key: 'handleMessage',
    value: _asyncToGenerator(function* (client, message) {
      var _this2 = this;

      var requestId = message.requestId;

      var returnVal = null;
      var returnType = { kind: 'void' };
      var callError = undefined;
      var hadError = false;

      try {
        switch (message.type) {
          case 'FunctionCall':
            // Transform arguments and call function.

            var _functionsByName$get = this._functionsByName.get(message['function']),
                fcLocalImplementation = _functionsByName$get.localImplementation,
                fcType = _functionsByName$get.type;

            var fcTransfomedArgs = yield Promise.all(message.args.map(function (arg, i) {
              return _this2._typeRegistry.unmarshal(arg, fcType.argumentTypes[i]);
            }));

            // Invoke function and return the results.
            returnType = fcType.returnType;
            returnVal = fcLocalImplementation.apply(this, fcTransfomedArgs);
            break;
          case 'MethodCall':
            // Get the object.
            var mcObject = this._objectRegistry.get(message.objectId);

            // Get the method FunctionType description.
            var mcType = this._classesByName.get(mcObject._interface).definition.instanceMethods.get(message.method);

            // Unmarshal arguments.
            var mcTransfomedArgs = yield Promise.all(message.args.map(function (arg, i) {
              return _this2._typeRegistry.unmarshal(arg, mcType.argumentTypes[i]);
            }));

            // Invoke message.
            returnType = mcType.returnType;
            returnVal = mcObject[message.method].apply(mcObject, mcTransfomedArgs);
            break;
          case 'NewObject':
            var _classesByName$get = this._classesByName.get(message['interface']),
                noLocalImplementation = _classesByName$get.localImplementation,
                noDefinition = _classesByName$get.definition;

            // Transform arguments.
            var noTransfomedArgs = yield Promise.all(message.args.map(function (arg, i) {
              return _this2._typeRegistry.unmarshal(arg, noDefinition.constructorArgs[i]);
            }));

            // Create a new object and put it in the registry.
            var noObject = construct(noLocalImplementation, noTransfomedArgs);

            // Return the object, which will automatically be converted to an id through the
            // marshalling system.
            returnType = { kind: 'promise', type: { kind: 'named', name: message['interface'] } };
            returnVal = Promise.resolve(noObject);
            break;
          case 'DisposeObject':
            // Get the object.
            var doObject = this._objectRegistry.get(message.objectId);

            // Remove the object from the registry, and scrub it's id.
            doObject._remoteId = undefined;
            this._objectRegistry['delete'](message.objectId);

            // Call the object's local dispose function.
            returnType = { kind: 'promise', type: { kind: 'void' } };
            yield doObject.dispose();

            // Return a void Promise
            returnVal = Promise.resolve();
            break;
          case 'DisposeObservable':
            // Dispose an in-progress observable, before it has naturally completed.
            if (this._subscriptions.has(requestId)) {
              this._subscriptions.get(requestId).dispose();
              this._subscriptions['delete'](requestId);
            }
            break;
          default:
            throw new Error('Unkown message type ' + message.type);
        }
      } catch (e) {
        logger.error(e != null ? e.message : e);
        callError = e;
        hadError = true;
      }

      switch (returnType.kind) {
        case 'void':
          break; // No need to send anything back to the user.
        case 'promise':
          // If there was an error executing the command, we send that back as a rejected promise.
          if (hadError) {
            returnVal = Promise.reject(callError);
          }

          // Ensure that the return value is a promise.
          if (!isThenable(returnVal)) {
            returnVal = Promise.reject(new Error('Expected a Promise, but the function returned something else.'));
          }

          // Marshal the result, to send over the network.
          returnVal = returnVal.then(function (value) {
            return _this2._typeRegistry.marshal(value, returnType.type);
          });

          // Send the result of the promise across the socket.
          returnVal.then(function (result) {
            var resultMessage = {
              channel: 'service_framework3_rpc',
              type: 'PromiseMessage',
              requestId: requestId,
              result: result,
              hadError: false
            };
            _this2._server._sendSocketMessage(client, resultMessage);
          }, function (error) {
            var errorMessage = {
              channel: 'service_framework3_rpc',
              type: 'ErrorMessage',
              requestId: requestId,
              hadError: true,
              error: formatError(error)
            };
            _this2._server._sendSocketMessage(client, errorMessage);
          });
          break;
        case 'observable':
          // If there was an error executing the command, we send that back as an error Observable.
          if (hadError) {
            returnVal = _rx.Observable['throw'](callError);
          }

          // Ensure that the return value is an observable.
          if (!isObservable(returnVal)) {
            returnVal = _rx.Observable['throw'](new Error('Expected an Observable, but the function returned something else.'));
          }

          // Marshal the result, to send over the network.
          returnVal = returnVal.concatMap(function (value) {
            return _this2._typeRegistry.marshal(value, returnType.type);
          });

          // Send the next, error, and completion events of the observable across the socket.
          var subscription = returnVal.subscribe(function (data) {
            var eventMessage = {
              channel: 'service_framework3_rpc',
              type: 'ObservableMessage',
              requestId: requestId,
              hadError: false,
              result: {
                type: 'next',
                data: data
              }
            };
            _this2._server._sendSocketMessage(client, eventMessage);
          }, function (error) {
            var errorMessage = {
              channel: 'service_framework3_rpc',
              type: 'ErrorMessage',
              requestId: requestId,
              hadError: true,
              error: formatError(error)
            };
            _this2._server._sendSocketMessage(client, errorMessage);
            _this2._subscriptions['delete'](requestId);
          }, function (completed) {
            var eventMessage = {
              channel: 'service_framework3_rpc',
              type: 'ObservableMessage',
              requestId: requestId,
              hadError: false,
              result: { type: 'completed' }
            };
            _this2._server._sendSocketMessage(client, eventMessage);
            _this2._subscriptions['delete'](requestId);
          });
          this._subscriptions.set(requestId, subscription);
          break;
        default:
          throw new Error('Unkown return type ' + returnType.kind + '.');
      }
    })
  }]);

  return ServerComponent;
})();

exports['default'] = ServerComponent;
function construct(classObject, args) {
  function F() {
    return classObject.apply(this, args);
  }
  F.prototype = classObject.prototype;
  return new F();
}

/**
 * A helper function that checks if an object is thenable (Promise-like).
 */
function isThenable(object) {
  return Boolean(object && object.then);
}

/**
 * A helper function that checks if an object is an Observable.
 */
function isObservable(object) {
  return Boolean(object && object.concatMap && object.subscribe);
}

/**
 * Format the error before sending over the web socket.
 * TODO: This should be a custom marshaller registered in the TypeRegistry
 */
function formatError(error) {
  if (error instanceof Error) {
    return {
      message: error.message,
      code: error.code,
      stack: error.stack
    };
  } else if (typeof error === 'string') {
    return error.toString();
  } else if (error === undefined) {
    return undefined;
  } else {
    return 'Unknown Error: ' + error.toString();
  }
}
module.exports = exports['default'];

/**
 * Store a mapping from function name to a structure holding both the local implementation and
 * the type definition of the function.
 */

/**
 * Store a mapping from a class name to a struct containing it's local constructor and it's
 * interface definition.
 */
