// nuclide-remote-connection/lib/main.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */
var RemoteConnection = require('./RemoteConnection');
var RemoteDirectory = require('./RemoteDirectory');
var RemoteFile = require('./RemoteFile');

var _require = require('./SshHandshake');

var SshHandshake = _require.SshHandshake;
var decorateSshConnectionDelegateWithTracking = _require.decorateSshConnectionDelegateWithTracking;

/*
 * We want the services to exist as a singleton. The ./service-manager.js file
 * in this package contains a cache of services, each of which should also be singletons for each
 * root/service pair. To this end, we use the singleton() utility in nuclide-commons to ensure that
 * only one instance of service-manager is loaded, serving as a clearinghouse for all service
 * loading, which is done via its getService() and getServiceByNuclideUri() methods.
 */
var NUCLIDE_CLIENT_EXPORTS_KEY = '_nuclide_client_exports';
var nuclideClientExports = require('nuclide-commons').singleton.get(NUCLIDE_CLIENT_EXPORTS_KEY, function () {
  var _require2 = require('./service-manager');

  var getService = _require2.getService;
  var getServiceByNuclideUri = _require2.getServiceByNuclideUri;
  var getServiceLogger = _require2.getServiceLogger;
  var getRemoteServiceByRemoteConnection = _require2.getRemoteServiceByRemoteConnection;

  var _require3 = require('./client');

  var getClient = _require3.getClient;
  var getFileForPath = _require3.getFileForPath;

  return {
    getClient: getClient,
    getFileForPath: getFileForPath,
    getService: getService,
    getServiceLogger: getServiceLogger,
    getServiceByNuclideUri: getServiceByNuclideUri,
    getRemoteServiceByRemoteConnection: getRemoteServiceByRemoteConnection
  };
});

module.exports = {
  decorateSshConnectionDelegateWithTracking: decorateSshConnectionDelegateWithTracking,
  RemoteConnection: RemoteConnection,
  RemoteFile: RemoteFile,
  RemoteDirectory: RemoteDirectory,
  SshHandshake: SshHandshake,
  getClient: nuclideClientExports.getClient,
  getFileForPath: nuclideClientExports.getFileForPath,
  getService: nuclideClientExports.getService,
  getServiceLogger: nuclideClientExports.getServiceLogger,
  getServiceByNuclideUri: nuclideClientExports.getServiceByNuclideUri,
  getRemoteServiceByRemoteConnection: nuclideClientExports.getRemoteServiceByRemoteConnection
};
